nexus/db-model/src/inventory.rs

```diff
@@ -881,16 +882,65 @@ impl InvPhysicalDisk {
     }
 }
 
-impl From<InvPhysicalDisk> for nexus_types::inventory::PhysicalDisk {
-    fn from(disk: InvPhysicalDisk) -> Self {
-        Self {
-            identity: omicron_common::disk::DiskIdentity {
-                vendor: disk.vendor,
-                serial: disk.serial,
-                model: disk.model,
-            },
-            variant: disk.variant.into(),
-            slot: disk.slot,
+/// See [`nexus_types::inventory::PhysicalDiskFirmware::Nvme`].
+#[derive(Queryable, Clone, Debug, Selectable, Insertable)]
+#[diesel(table_name = inv_nvme_disk_firmware)]
+pub struct InvNvmeDiskFirmware {
+    pub inv_collection_id: DbTypedUuid<CollectionKind>,
+    pub sled_id: DbTypedUuid<SledKind>,
+    pub slot: i64,
+    pub active_slot: SqlU8,
+    pub next_active_slot: Option<SqlU8>,
+    pub number_of_slots: SqlU8,
+    pub slot1_is_read_only: bool,
+    pub slot_firmware_versions: Vec<Option<String>>,
+}
+
+impl InvNvmeDiskFirmware {
+    pub fn try_from_physical_disk(
+        inv_collection_id: CollectionUuid,
+        sled_id: SledUuid,
+        disk: nexus_types::inventory::PhysicalDisk,
+    ) -> Option<Self> {
+        match disk.firmware {
+            PhysicalDiskFirmware::Unknown => None,
+            PhysicalDiskFirmware::Nvme(firmware) => Some(Self {
+                inv_collection_id: inv_collection_id.into(),
+                sled_id: sled_id.into(),
+                slot: disk.slot,
+                active_slot: firmware.active_slot.into(),
+                next_active_slot: firmware
+                    .next_active_slot
+                    .map(|nas| nas.into()),
+                number_of_slots: firmware.number_of_slots.into(),
+                slot1_is_read_only: firmware.slot1_is_read_only,
+                slot_firmware_versions: firmware.slot_firmware_versions,
+            }),
+        }
+    }
+
+    /// Attempt to read the current firmware version.
```

Comment:
There's a lot of logic around these slots / next versions, but I don't really see any explicit tests for this. I'd be interested in seeing:

- Having current versions / next versions at the minimum / maximum values
- Attempting to set a "slot" of zero, 8, observing that the system doesn't crash

I'd be particularly interested in seeing these tests actually try to put these values into the DB, because we also have DB constraints that should be validated.